### PR TITLE
[exporter/elasticsearch] Add initial support for metrics

### PR DESCRIPTION
**Pull request by andrzej-stencel**
**Description:**

Adds initial support for metrics. Currently only Sum ang Gauge metric types are supported, other metric types are ignored.

**Testing:**

Added unit tests.

**Documentation:**

Added docs in component README.

### CODE REVIEW

File: exporter/elasticsearchexporter/attribute.go [hunk 1]
```diff
@@ -37,3 +38,12 @@ func getFromAttributes(name string, resource, scope, record attrGetter) string {
 	}
 	return str
 }
+
+func getFromAttributesNew(name string, defaultValue string, attributeMaps ...pcommon.Map) string {
```
Comment (andrzej-stencel) [inline exporter/elasticsearchexporter/attribute.go:42]: I want to create a separate pull request where I replace the deprecated `getFromAttributes` function with the new one.

File: exporter/elasticsearchexporter/exporter.go [hunk 2]
```diff
@@ -132,11 +133,69 @@ func (e *elasticsearchExporter) pushLogRecord(ctx context.Context, resource pcom
...
+	ctx context.Context,
+	metrics pmetric.Metrics,
+) error {
+	var errs []error
+
+	rls := metrics.ResourceMetrics()
+	for i := 0; i < rls.Len(); i++ {
+		rl := rls.At(i)
+		resource := rl.Resource()
+		ills := rl.ScopeMetrics()
```
Comment (axw) [inline exporter/elasticsearchexporter/exporter.go:151]: nit: not clear what "ills" is (is this copied from logs? instruction scope logs?), could we give it a more meaningful name such as "scopeMetrics"?

File: exporter/elasticsearchexporter/model.go [hunk 3]
```diff
@@ -155,6 +161,131 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
 	return document
 }
 
+func (m *encodeModel) encodeMetrics(resource pcommon.Resource, metrics pmetric.MetricSlice, _ pcommon.InstrumentationScope) ([][]byte, error) {
+	hasher := sha256.New()
```
Comment (axw) [inline exporter/elasticsearchexporter/model.go:165]: Is using a cryptographic hash necessary? I think we should probably use a fast, non-cryptographic hash algorithm like xxhash (specifically github.com/cespare/xxhash/v2)

File: exporter/elasticsearchexporter/model.go [hunk 4]
```diff
@@ -155,6 +161,131 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
...
+			doc.Dedup()
+		} else if m.dedot {
+			doc.Sort()
+		}
+
+		var buf bytes.Buffer
+		err := doc.Serialize(&buf, m.dedot)
+
+		if err != nil {
+			fmt.Printf("Serialize error, dropping doc: %v\n", err)
```
Comment (axw) [inline exporter/elasticsearchexporter/model.go:220]: Should we be returning the error?

File: exporter/elasticsearchexporter/model.go [hunk 5]
```diff
@@ -155,6 +161,131 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
...
+
+func metricHash(hasher hash.Hash, timestamp pcommon.Timestamp, attributes pcommon.Map) string {
+	// Avoid any hashing if there are no attributes
+	if attributes.Len() == 0 {
+		return timestamp.String()
+	}
+
+	hasher.Reset()
+
+	hasher.Write([]byte(timestamp.AsTime().Format(time.RFC3339Nano)))
```
Comment (axw) [inline exporter/elasticsearchexporter/model.go:237]: Encoding the uint64 bytes should avoid some allocs, and generally be cheaper:

```go
var buf [8]byte
binary.LittleEndian.PutUint64(buf[:], uint64(timestamp))
hash.Write(buf[:])
```

File: exporter/elasticsearchexporter/model.go [hunk 6]
```diff
@@ -155,6 +161,131 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
...
+	mapHash(hasher, attributes)
+
+	return string(hasher.Sum(nil))
+}
+
+func mapHash(hasher hash.Hash, m pcommon.Map) {
+	// TODO: Only hasing values, not keys? :thinking:
+	m.Range(func(_ string, v pcommon.Value) bool {
+		hasher.Write([]byte(v.Str()))
+		valueHash(hasher, v)
```
Comment (axw) [inline exporter/elasticsearchexporter/model.go:248]: Yes, I do think we want to hash the keys as well.
```suggestion
	m.Range(func(k string, v pcommon.Value) bool {
		hasher.Write([]byte(k.Str()))
		valueHash(hasher, v)
```

**Review by axw — COMMENTED** [review]
Unsolicited review, hope you don't mind :) Looks great, just a few minor comments.

File: exporter/elasticsearchexporter/exporter.go (hunk 2 above)
Comment (andrzej-stencel) [reply exporter/elasticsearchexporter/exporter.go]: Good point, I actually hate it when variable names don't make it easy to understand the code. Guilty as charged!
Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33513/commits/1136e74c2d4a675f42c24b2b65c09873367711d4.

File: exporter/elasticsearchexporter/model.go (hunk 3 above)
Comment (andrzej-stencel) [reply exporter/elasticsearchexporter/model.go:165]: Very good point. This reminds me that [the OTel Collector contributing guide](https://github.com/open-telemetry/opentelemetry-collector/blob/v0.102.1/CONTRIBUTING.md#recommended-libraries--defaults) specifically recommends [hash/fnv](https://pkg.go.dev/hash/fnv).

Comment (andrzej-stencel) [reply exporter/elasticsearchexporter/model.go]: Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33513/commits/cf4d392c28e6f67ba9cbee80fa211f887ff11a11.

File: exporter/elasticsearchexporter/model.go (hunk 5 above)
Comment (andrzej-stencel) [reply exporter/elasticsearchexporter/model.go]: Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33513/commits/cf4d392c28e6f67ba9cbee80fa211f887ff11a11.

File: exporter/elasticsearchexporter/model.go (hunk 6 above)
Comment (andrzej-stencel) [reply exporter/elasticsearchexporter/model.go]: Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33513/commits/cf4d392c28e6f67ba9cbee80fa211f887ff11a11.

File: exporter/elasticsearchexporter/model.go (hunk 4 above)
Comment (andrzej-stencel) [reply exporter/elasticsearchexporter/model.go]: I changed this code to return the error: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33513/commits/2e9cfb82aef46db9d5bbaf6c514e0cca39b2fc00.

**Comment by andrzej-stencel** [comment]
> Unsolicited review, hope you don't mind :) Looks great, just a few minor comments.

Thank you so much @axw for the review, your comments were very helpful! Keep them coming! :pray:

File: exporter/elasticsearchexporter/model.go [hunk 7]
```diff
@@ -155,6 +161,126 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
...
+	case pcommon.ValueTypeSlice:
+		sliceHash(h, v.Slice())
+	}
+}
+
+func sliceHash(h hash.Hash, s pcommon.Slice) {
+	for i := 0; i < s.Len(); i++ {
+		valueHash(h, s.At(i))
+	}
+}
```
Comment (axw) [inline exporter/elasticsearchexporter/model.go:282]: I just came across https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/pdatautil/hash.go, which does a lot of this already (but using xxhash). Perhaps we should extend and use that?

I think the main thing we would be missing from pdatautil is the ability to keep state while hashing multiple values -- timestamp & attributes in our case. We could alternatively key on (timestamp, hash(attributes)).

Lastly, we could alternatively build on https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/exp/metrics/identity (specifically "OfStream"). But again, that doesn't include timestamp...

**Review by axw — APPROVED** [review]
LGTM. I'm mildly concerned about using hash/fnv, particularly 32-bit, without any chaining for handling collisions. There's some existing code we can build on, but I think it would be reasonable to defer that to a followup.

File: exporter/elasticsearchexporter/model.go (hunk 7 above)
Comment (andrzej-stencel) [reply exporter/elasticsearchexporter/model.go:282]: Thanks for finding pdatautil, I looked briefly for something like this but didn't see it. Definitely worth looking into it.

File: exporter/elasticsearchexporter/model_test.go [hunk 8]
```diff
@@ -22,6 +26,23 @@ var expectedSpanBody = `{"@timestamp":"2023-04-19T03:04:05.000000006Z","Attribut
 
 var expectedLogBody = `{"@timestamp":"2023-04-19T03:04:05.000000006Z","Attributes.log-attr1":"value1","Body":"log-body","Resource.key1":"value1","Scope.name":"","Scope.version":"","SeverityNumber":0,"TraceFlags":0}`
 
+var expectedMetricsEncoded = `{"@timestamp":"2024-06-12T10:20:16.419290690Z","cpu":"cpu0","host":{"name":"my-host"},"os":{"type":"linux"},"state":"idle","system":{"cpu":{"time":440.23}}}
+{"@timestamp":"2024-06-12T10:20:16.419290690Z","cpu":"cpu0","host":{"name":"my-host"},"os":{"type":"linux"},"state":"interrupt","system":{"cpu":{"time":0}}}
```
Comment (felixbarny) [inline exporter/elasticsearchexporter/model_test.go:30]:
```suggestion
{"@timestamp":"2024-06-12T10:20:16.419290690Z","cpu":"cpu0","host":{"name":"my-host"},"os":{"type":"linux"},"state":"interrupt","system":{"cpu":{"time":0.0}}}
```

Let's try to force a fractional to be present for floating point numbers, so that ES can correctly do dynamic mappings.

Comment (andrzej-stencel) [reply exporter/elasticsearchexporter/model_test.go:30]: Perhaps we can do this in a follow-up PR?

Comment (felixbarny) [reply exporter/elasticsearchexporter/model_test.go:30]: sounds good

File: exporter/elasticsearchexporter/model.go (hunk 7 above)
Comment (andrzej-stencel) [reply exporter/elasticsearchexporter/model.go:289]: Due to a change in the implementation, the whole hashing part is now  removed.

File: exporter/elasticsearchexporter/model.go [hunk 9]
```diff
@@ -163,6 +169,125 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
...
+			dps = metric.Sum().DataPoints()
+		}
+
+		for j := 0; j < dps.Len(); j++ {
+			dp := dps.At(j)
+
+			hash := metricHash(dp.Timestamp(), dp.Attributes())
+			doc, docExists := docs[hash]
+			if !docExists {
+				doc = baseDoc.Clone()
```
Comment (carsonip) [inline exporter/elasticsearchexporter/model.go:198]: IIUC `Clone` is not a deep clone, and dedup is potentially going to mutate the field. I feel more comfortable with a deep clone. In any case, can you add a test where in resource attributes and dp attributes, there are duplicate fields, as well as renamed fields (see Dedup logic), and ensure that the output is still expected?

File: exporter/elasticsearchexporter/model.go [hunk 10]
```diff
@@ -163,6 +169,125 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
...
+			dps = metric.Gauge().DataPoints()
+		case pmetric.MetricTypeSum:
+			dps = metric.Sum().DataPoints()
+		}
+
+		for j := 0; j < dps.Len(); j++ {
+			dp := dps.At(j)
+
+			hash := metricHash(dp.Timestamp(), dp.Attributes())
+			doc, docExists := docs[hash]
```
Comment (carsonip) [inline exporter/elasticsearchexporter/model.go:196]: non-blocker: Are we worried about hash collisions? In apm-data it doesn't just compare hashes, but compares string-based map keys, which in go implementation should first compare hash then actual bytes (see [code](https://github.com/elastic/apm-data/blob/main/input/otlp/metrics.go#L378)). However, at the same time, apm-data is vulnerable to value->string collisions.

File: .chloggen/add-metrics-to-elasticsearch-exporter.yaml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+subtext:
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: []
```
[the remark this answers is not in the document]
Comment (andrzej-stencel) [reply .chloggen/add-metrics-to-elasticsearch-exporter.yaml:27]: The feature is available to users, so yes, it is relevant. `user` is the default value for `change_logs`, so no need to specify it explicitly.

File: exporter/elasticsearchexporter/model.go (hunk 10 above)
Comment (andrzej-stencel) [reply exporter/elasticsearchexporter/model.go:196]: Oh this whole hashing detail will actually be removed in a follow up PR ([draft here](https://github.com/andrzej-stencel/opentelemetry-collector-contrib/pull/1)) due to a change in how data is routed. Sorry for the trouble, probably no need reviewing this in light of the upcoming change.

File: .chloggen/add-metrics-to-elasticsearch-exporter.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (carsonip) [reply .chloggen/add-metrics-to-elasticsearch-exporter.yaml:27]: TIL!

**Review by carsonip — APPROVED** [review]

**Review by carsonip — APPROVED** [review]

**Review by mx-psi — APPROVED** [review]
